paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc

```diff
@@ -61,414 +61,428 @@ static mkldnn::memory::data_type GetDstType(bool is_int8, bool is_bfloat16,
   return dst_dt;
 }
 
+template <typename T>
+struct remap {
+  // Default: Output type is the same as input type.
+  typedef T type;
+};
+
+template <>
+struct remap<uint8_t> {
+  typedef int8_t type;
+};
+
+template <typename T>
+std::shared_ptr<std::tuple<float, std::vector<float>>> get_bias_scales(
+    const framework::ExecutionContext& ctx,
+    const platform::MKLDNNDeviceContext& dev_ctx, const std::string& key) {
+  return std::make_shared<std::tuple<float, std::vector<float>>>(
+      std::make_tuple(0.0f, std::vector<float>(1, 1.0f)));
+}
+
+template <>
+std::shared_ptr<std::tuple<float, std::vector<float>>> get_bias_scales<int8_t>(
+    const framework::ExecutionContext& ctx,
+    const platform::MKLDNNDeviceContext& dev_ctx, const std::string& key) {
+  // Get scales int8 bias key
+  const std::string key_bs = key + "@bs";
+
+  // Scales for int8 bias are to be cached to avoid
+  // computing them each iteration
+  auto bias_scale_tuple =
+      std::static_pointer_cast<std::tuple<float, std::vector<float>>>(
+          dev_ctx.GetBlob(key_bs));
+  if (bias_scale_tuple) return bias_scale_tuple;
+
+  const auto* filter = ctx.Input<Tensor>("Filter");
+  const auto& weights_tz = framework::vectorize(filter->dims());
+  const int groups = std::max(ctx.Attr<int>("groups"), 1);
+
+  const auto& scale_weights_data =
+      ctx.Attr<std::vector<float>>("Scale_weights");
+  const auto& scale_in_data = ctx.Attr<float>("Scale_in");
+
+  bool is_multi_channel = scale_weights_data.size() > 1;
+  int mask_reorder = is_multi_channel ? 1 << 0 : 1;
+  const int count =
+      is_multi_channel
+          ? (groups > 1 ? (weights_tz)[1] * (weights_tz)[0] : (weights_tz)[0])
+          : 1;
+
+  bias_scale_tuple =
+      std::make_shared<std::tuple<float, std::vector<float>>>(std::make_tuple(
+          static_cast<float>(mask_reorder), std::vector<float>(count)));
+  for (int i = 0; i < count; i++) {
+    std::get<1>(*bias_scale_tuple)[i] = scale_in_data * scale_weights_data[i];
+  }
+
+  dev_ctx.SetBlob(key_bs, bias_scale_tuple);
+
+  return bias_scale_tuple;
+}
```

Comment:
Maybe merging these 2 functions into one and doing something different depending on the result of:
`if(std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value`
would be a better solution?
It would allow to do get rid of the remap structs that seemed quite confusing to me at first glance.

Reply:
I have diffrent observation e.g. I think remapping is way easier for me to understand:) And also "if.." is a runtime check as c++14 does not have if contrexpr (c++17) , and remapping mechanism makes a check evaluated in compile time